lib/ex_unit/test/ex_unit/pattern_format_test.exs

```diff
@@ -0,0 +1,581 @@
+      "}",
+      ", ",
+      "e: ",
+      "\"hello\"",
+      "}"
+    ]
+
+    assert actual == expected
+  end
+
+  test "Large, nested map, missing" do
+    # send(self(), %{a: 1, b: %{a: 2, c: %{d: 4}}, e: "hello"})
+    # assert_receive %{a: 1, b: %{a: 2, c: %{d: 4, f: "world"}}, e: "hello"}
+
+    left = quote do: %{a: 1, b: %{a: 2, c: %{d: 4, f: "world"}}, e: "hello"}
+    pattern = Pattern.new(left, [], %{})
+
+    actual = PatternFormat.script(pattern, %{a: 1, b: %{a: 2, c: %{d: 4}}, e: "hello"})
+
+    expected = [
+      "%{",
+      "a: ",
+      "1",
+      ", ",
+      "b: ",
+      "%{",
+      "a: ",
+      "2",
+      ", ",
+      "c: ",
+      "%{",
+      "d: ",
+      "4",
+      ", ",
+      {:diff_insert, "f: \"world\""},
+      "}",
+      "}",
+      ", ",
+      "e: ",
+      "\"hello\"",
+      "}"
+    ]
+
+    assert actual == expected
+  end
+
+  test "Large, nested map, extra" do
+    # this test passes
+    # send(self(), %{a: 1, b: %{a: 2, c: %{d: 4, f: "world"}}, e: "hello"})
+    # assert_receive %{a: 1, b: %{a: 2, c: %{d: 4}}, e: "hello"}
+
+    left = quote do: %{a: 1, b: %{a: 2, c: %{d: 4}}, e: "hello"}
+
+    pattern = Pattern.new(left, [], %{})
+
+    actual =
+      PatternFormat.script(pattern, %{a: 1, b: %{a: 2, c: %{d: 4, f: "world"}}, e: "hello"})
+
+    expected = [
+      "%{",
+      "a: ",
+      "1",
+      ", ",
+      "b: ",
+      "%{",
+      "a: ",
+      "2",
+      ", ",
+      "c: ",
+      "%{",
+      "d: ",
+      "4",
+      ", ",
+      "f: \"world\"",
+      "}",
+      "}",
+      ", ",
+      "e: ",
+      "\"hello\"",
+      "}"
+    ]
+
+    assert actual == expected
+  end
+
+  test "using variables" do
+    # send(self(), [1, 2, 3])
+    # assert_receive([a, a, 3])
+
+    left = quote do: [a, a, 3]
+    pattern = Pattern.new(left, [], %{a: :ex_unit_unbound_var})
+
+    actual = PatternFormat.script(pattern, [1, 2, 3])
+
+    expected = ["[", "1", ", ", {:diff_delete, "2"}, ", ", "3", "]"]
+    assert actual == expected
+  end
+
+  test "using guards" do
+    # send(self(), [1, 2, 3])
+    # assert_receive [a, 2, 3] when is_binary(a)
+
+    left = quote do: [a, 2, 3] when is_binary(a)
+
+    pattern = Pattern.new(left, [], %{a: :ex_unit_unbound_var})
+
+    actual = PatternFormat.script(pattern, [1, 2, 3])
+
+    expected = ["[", "1", ", ", "2", ", ", "3", "]", " ", {:diff_delete, "when is_binary(a)"}]
+    assert actual == expected
+  end
+
+  test "using guards, but guard matches" do
+    # send(self(), [1, 2, 3])
+    # assert_receive [a, b, 3, 4] when is_integer(a)
+
+    left = quote do: [a, b, 3, 4] when is_binary(a)
+
+    pattern = Pattern.new(left, [], %{a: :ex_unit_unbound_var, b: :ex_unit_unbound_var})
+
+    actual = PatternFormat.script(pattern, [1, 2, 3])
+
+    expected = [
+      "[",
+      "1",
+      ", ",
+      "2",
+      ", ",
+      "3",
+      ", ",
+      {:diff_insert, "4"},
+      "]",
+      " ",
+      {:diff_delete, "when is_binary(a)"}
```

Comment:
This is interesting. We are showing the guard but we don't have a variable. I am wondering if this can be confusing or not.

Another approach would be to show the diff on the pattern. But this complicates things for `assert_receive`. I.e. we would need to show the pattern alongside each attempted message.

What are your thoughts between the approaches?

Reply:
I went back and forth between the variable, and value, and I can be easily talked into either.   I think showing `when is_binary(2)` in red actually makes it easier to see what the problem is, but `when is_binary(a)` points directly to the pattern - and highlights that this is being brought in from the pattern, and not a part of the item being matched.  What are your preferences?

Reply:
I think was not clear enough.I was thinking about a different approach where the diff actually applies to the pattern. So in such cases, we would show the pattern using green/red and not the value. I can see some advantages of this approach:

For example, on a match like:

    %{foo: _, bar: _} = really_large_map

where the bar key is missing, it would show foo in green, bar in red, so it would be clearer that the bar key is missing. If we had a large map on the right side, we can't show misses.

This approach would also work better with guards.

This approach can also be useful in other situations, for example, in the `UndefinedFunctionClauseError`, we could perform a better diff there too, as we show the patterns there.

The only problem I can think is regarding `assert_receive`, where we would need to show the pattern alongside each value... but maybe it is not that bad?

Thoughts?

Reply:
Maybe more of an `and` than an `or`?  Colored pattern on the left, colored map on the right?  I think it would just be screen spam for assert_receive if there were more than couple of messages in the mailbox.  However, we now have a mechanism to determine how closely a messages matches, so we could zero in on a one  'best choices'?

Reply:
I am worried about the complexity of supporting both approaches. But I am not sure how to handle `assert_receive`. Maybe we should try some "mock ups" of how the CLI would look like? I can send something soon. :D

Reply:
Here are some mockups:

![screen shot 2018-10-10 at 01 40 22](https://user-images.githubusercontent.com/651203/46704982-c7a6eb80-cc2d-11e8-97e2-f19e6a7ce222.png)

I'll work tomorrow on the ones for `assert_receive` 🙂

Reply:
Thanks @fertapric! Were those mockups generated by hand or are you running this PR?

Reply:
By hand. 

I think we can use them to discuss the user experience of the CLI, and then check the implementation.

This is what I used: https://gist.github.com/fertapric/c7ee41e890214b1123e76f17dfd288c4. The code couldn't be uglier 🙈, but it works!

Reply:
I assume that the 'code:', 'left:', 'right:' and 'stacktrace:' are supposed to be in the existing blue, and not green as shown here?   I have a slight concern that this depicts symmetrically sized left & right, and in practice, we generally resort to asserting a pattern match primarily for large, nested structs, so there is almost always asymmetry in size between the pattern on the left (a short value), and the target on the right (often a very long value).   But I think that what is detailed here will work well.